pynumaflow/function/_dtypes.py

```diff
@@ -13,32 +13,61 @@
 Ms = TypeVar("Ms", bound="Messages")
 MT = TypeVar("MT", bound="MessageT")
 MTs = TypeVar("MTs", bound="MessageTs")
+N = TypeVar("N", bound="NoPublicConstructor")
+
+
+class NoPublicConstructor(type):
+    """Metaclass that ensures a private constructor
+
+    If a class uses this metaclass like this:
+
+        class SomeClass(metaclass=NoPublicConstructor):
+            pass
+
+    If you try to instantiate your class using (`SomeClass()`), a `TypeError` will be thrown.
+    """
+
+    def __call__(cls, *args, **kwargs):
+        raise TypeError(
```

Comment:
Can you add a custom exception here? e.g. `NoPublicConstructorError`

Reply:
Good idea, addressed. +1